xpublish/rest.py

```diff
@@ -106,8 +114,86 @@ class Rest:
 
     """
 
-    def __init__(self, datasets, routers=None, cache_kws=None, app_kws=None):
+    def __init__(
+        self,
+        datasets,
+        routers=None,
+        cache_kws=None,
+        app_kws=None,
+        plugins: Optional[Dict[str, XpublishPluginFactory]] = None,
+        extend_plugins: Optional[Dict[str, XpublishPluginFactory]] = None,
+        exclude_plugin_names: Optional[List[str]] = None,
+        plugin_configs: Optional[Dict] = None,
+    ):
+
+        dataset_route_prefix = self.init_datasets(datasets)
+
+        if plugins is not None:
+            self._plugins = plugins
+        else:
+            self.load_plugins(exclude_plugins=exclude_plugin_names, plugin_configs=plugin_configs)
```

Comment:
I started with a simpler approach, but then I tried to put myself in the position of a bunch of different user cases:
- 'I'm happy to use all the defaults'
- 'I built my own plugin! How do I add that!'
- 'My org has said that I can't use X due to inane security reasons..., how can I disable just that and keep the rest'
- 'I like most of what comes by default, but I need to pass a little configuration to X plugin'

I think the differences and reasons why each is useful would probably be made clear with some docs, tests, and examples. 

```python
rest = xpublish.Rest(
    datasets,
    routers=[
        (dap_router, {"tags": ["opendap"], "prefix": "/opendap"}),  # manually adding a router that isn't available as a plugin
    ],
    # plugins={}, # don't override plugins, we want to largely load what's available automatically
    plugin_configs={
        # "cf_edr": {
        #     "dataset_router_prefix": "/cf_edr_test_override"  # passing config info into a plugin loaded via entry points
        # }
    },
    extend_plugins={
        "cf_edr": CfEdrPlugin(dataset_router_prefix="/cf_edr_extended") # adding a new plugin, or overriding one that was loaded via entry points
    },
    exclude_plugin_names=["module_version"]  # don't share my dirty laundry
)
```

Also my naming of things isn't the clearest in the flurry of hacking.

Reply:
I also find all those use cases relevant and useful! However, the API is still a bit confusing to me. I'd suggest put it out of `Rest`. Also maybe accept both simple fastapi routers and plugins via a single argument (I agree it nice to keep a simple way to define dataset API routes). 

A rough sketch:

```python
T_Router = Union[XpublishPluginFactory, tuple[fastapi.APIRouter, dict[str, Any]]]


def load_default_routers(
    exclude: iterable[str] | None = None,
) -> dict[str, T_Router]:
    """return a dict of router plugins loaded from entrypoints."""
    return ...


class Rest:

    def __init__(
        self,
        datasets: Mapping[str, xr.Dataset],
        routers: Mapping[str, T_Router] | None = None,
    ):
        if routers is None:
            routers = load_default_routers()
        
        ...

#
# Usage examples
#

# use all routers but "module_version"
routers = load_default_routers(exclude=["module_version"])

# re-configure the "cf_edr" (entrypoint) router
routers["cf_edr"].configure(dataset_router_prefix="/cf_edr_test_override")

# or replace it completely
routers["cf_edr"] = CfEdrPlugin(dataset_router_prefix="/cf_edr_extended")

# add a simple router
routers["dap"] = (dap_router, {"tags": ["opendap"], "prefix": "/opendap"})

rest = Rest({...}, routers=routers)
```

Reply:
> I can't use X due to inane security reasons

(idea for later: we could maybe add a specific tag for plugins so that they are automatically selected or excluded in production vs. stagging vs. debug deployment when calling `load_default_routers`).

Reply:
Hmm, I think combining plugins and (dataset) routers might be a bit confusing, especially since plugins may eventually be more than just routers, how about something in between? 

Routers and plugins as init kwargs, and some helpers for default plugins?

```py
def load_default_plugins(
    exclude: iterable[str] | None = None,
) -> dict[str, XpublishPluginFactory]:
    """return a dict of plugins loaded from entrypoints."""
    return ...


class Rest:

    def __init__(
        self,
        datasets: Mapping[str, xr.Dataset],
        dataset_routers: Iterable[tuple[fastapi.APIRouter, dict[str, Any]]] | None = None,
        plugins: Mapping[str, XpublishPluginFactory] | None = None
    ):
        if plugins is None:
            plugins = load_default_plugins()
        
        ...

#
# Usage examples
#

# use all routers but "module_version"
plugins = load_default_plugins(exclude=["module_version"])

# re-configure the "cf_edr" (entrypoint) router
plugins["cf_edr"].configure(dataset_router_prefix="/cf_edr_test_override")

# or replace it completely
plugins["cf_edr"] = CfEdrPlugin(dataset_router_prefix="/cf_edr_extended")

# add a simple router
routers = [(dap_router, {"tags": ["opendap"], "prefix": "/opendap"})]

rest = Rest({...}, dataset_routers=routers, plugins=plugins)
```

Probably would wrap the helper in a method, so it's easier to override just that in a subclass.

> > I can't use X due to inane security reasons
> 
> (idea for later: we could maybe add a specific tag for plugins so that they are automatically selected or excluded in production vs. stagging vs. debug deployment when calling `load_default_routers`).

Yes, but at what point are we pulling the classic Flask move of reinventing Django?

Reply:
> Yes, but at what point are we pulling the classic Flask move of reinventing Django?

There's a long way before even approaching that point, but yeah we probably down want to go deep down that way. (just put some random thoughts for later).

Reply:
>  I think combining plugins and (dataset) routers might be a bit confusing, especially since plugins may eventually be more than just routers, how about something in between?

Hmm I'm not sure for what else could be reused the plugin base class and entry point added in this PR. Do you have some examples in mind?

Reply:
Ah, so I was thinking that this was only the start of a plugin interface, not the end point.

Some ideas:
- Dataset loaders
- Django debug style toolbar
- Dataset transforming middleware
- CLI tools

Reply:
I was pondering this as I was trying to think of the entry point group, but I think as soon as we try to split different functionality up into different types of plugins, we'll immediately have use cases that require multiple types to work in conjunction. 

I had another example in mind, but I've forgotten it, but at the moment what I can think of is a database backed dataset catalog. It may need to define it's own top level routes, override `get_dataset` and similar dependencies, and probably have CLI commands for managing the database structure. While those could possibly be three different plugins that get the same config.

What do you think about having a single top level Plugin class, and then nested specialty classes? (maybe use `pydantic.BaseModel` to help make it easier to set nested config)

```py
class XpublishPlugin(BaseModel):
    app_router: Router | None
    dataset_router: Router | None
    dependency_overrides: ...
    cli: ...
    dataset_middleware: ...

class Extension(BaseModel):
    plugin: XpublishPlugin

class Router(Extension):
    default_prefix: str
    ...
```

I think in most cases we would see 1 plugin for 1 package, but there are definitely cases where a package might expose multiple plugins. 

For example from the OGC EDR one that I've made, it might register a plugin that uses CF conventions by default, but have a variant that could be used if all the datasets aren't CF compliant and require a bit more configuration.